### PR TITLE
Free boundary conditions

### PR DESCRIPTION
**Pull request by abadams**
This branch replaces specialize_clamped_ramps and
specialize_branched_loops with a single simpler pass that forks loops into
(up to) three: a prologue, a simplified steady-state, and an epilogue.

Parallel loops are converted to if statements for the steady-state and
not-steady-state paths.

specialize_branched_loops had a tough combinatorial problem trying to
figure out which branches to make given a soup of generally
incomparable conditions. This pass requires that the steady state is
tagged by the programmer as the 'likely' one, via an intrinsic. Most
commonly this happens invisibly just by using the boundary condition
helpers. A few other things are marked as likely by Halide internally.

- The warm-up code in a sliding window optimization is pulled out as a prologue

- The last instance of a split that gets shifted leftwards is pulled outside of the loop.
This simplifies the inner loop code some for simple things like vectorized memcpy
by avoiding the max operation. It also means you can guarantee alignment in some
situations where you couldn't before.

- Clamped ramps are considered likely to simplify to the ramp. This is
done to help people who just use clamp directly instead of one of the
boundary condition helpers.

The main effect of this branch is better performance for boundary conditions.
Running make performance_boundary_conditions:

Trunk:

constant_exterior: 8.534000
repeat_edge: 3.061000
repeat_image: 73.329000
mirror_image: 87.948000
mirror_interior: 76.945000
unbounded: 3.219000

This branch:

constant_exterior: 2.807000
repeat_edge: 2.427000
repeat_image: 2.973000
mirror_image: 3.280000
mirror_interior: 3.174000
unbounded: 2.600000

A secondary effect is that you can now iterate over non-square
domains, by making the unlikely path a no-op that gets simplified
away. See test/correctness/iterate_over_circle.cpp:

    Expr t = cast<int>(ceil(sqrt(10*10 - y*y)));
    f(x, y) += select(x > -t && x < t, likely(in(x, y)), 0);

See test/correctness/likely.cpp for more examples.

I've found no huge performance differences in any code I have access to. I think people avoided using the fancier boundary conditions, or worked around performance issues with them. Some things get mildly faster. There have been no statistically significant regressions.

### CODE REVIEW

**Comment by steven-johnson** [comment]
Sounds like this makes proper use of BoundaryConditions a performance win now (rather than just a clarity win); if so, we should probably ensure that the docs make this clear and encourage it, and also that all our sample code and tutorials prefer it (rather than the older pattern of "throw clamps on the input").

File: src/PartitionLoops.cpp [hunk 1]
```diff
@@ -0,0 +1,926 @@
...
+
+            if (op->for_type == ForType::Serial) {
+                // Steady state.
+                new_loop = For::make(op->name, min_steady, max_steady - min_steady,
+                                     op->for_type, op->device_api, simpler_body);
+
+                if (make_prologue) {
+                    Stmt prologue = For::make(op->name, op->min, min_steady - op->min,
+                                              op->for_type, op->device_api, body);
+                    //prologue = simplify_prologue.mutate(prologue);
```
Comment (kree-colemcalughlin) [inline src/PartitionLoops.cpp]: Why is this commented out? It looks like you are not applying the bounds simplification to either the prologue or the epilogue right now.

File: src/PartitionLoops.cpp [hunk 2]
```diff
@@ -0,0 +1,926 @@
...
+
+class PartitionLoops : public IRMutator {
+    using IRMutator::visit;
+
+    void visit(const For *op) {
+
+        Stmt body = mutate(op->body);
+
+        // Conservatively only apply this optimization at one loop
+        // level. Comment this out to allow full expansion.
```
Comment (kree-colemcalughlin) [inline src/PartitionLoops.cpp]: If I understand correctly this implies that we will only get the performance win for boundary conditions in one dimension  (the innermost dim, I guess). Are we just worried about combinatorial explosion if we apply it to all loop levels? Could we maybe introduce a bound > 1 for the number of levels to apply this to?

File: src/Solve.cpp [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+            expr = op;
+        } else if (scope.contains(op->name)) {
+            CacheEntry e = scope.get(op->name);
+            expr = e.expr;
+            uses_var = uses_var || e.uses_var;
+        } else if (external_scope.contains(op->name)) {
+            Expr e = external_scope.get(op->name);
+            // Expressions in the external scope haven't been solved
+            // yet.
+            expr = mutate(e);
```
Comment (kree-colemcalughlin) [inline src/Solve.cpp]: Why don't we need to cache results from the external scope? Couldn't this class receive a external scope that would lead to combinatorial problems?

**Comment by kree-colemcalughlin** [comment]
This PR is great! I think the iterations over non-square domains is actually one of the coolest things here, and should be highlighted. I'd also +1 Steven's comments about making sure we emphasize that using the boundary conditions is the right way now.

So maybe making sure that all the apps and tutorials are using boundary conditions rather than clamps, would be an important addition to this or an accompanying PR. At the very least we should make sure we have some tutorials highlighting the use of boundary conditions.

Other than the few questions/comments I posted it LGTM.

File: src/BoundaryConditions.cpp [hunk 4]
```diff
@@ -4,59 +4,55 @@ namespace Halide {
 
 namespace BoundaryConditions {
 
-Func constant_exterior(const Func &source, Expr value,
-                       const std::vector<std::pair<Expr, Expr> > &bounds) {
+Func repeat_edge(const Func &source,
```
Comment (steven-johnson) [inline src/BoundaryConditions.cpp]: Did repeat_edge and constant_exterior just get re-ordered? It's a little hard to tell from this diff. (If it's just a re-order, perhaps putting them back in the same order for this PR would simplify review.)

File: src/IROperator.h [hunk 5]
```diff
@@ -1404,6 +1404,24 @@ inline Expr memoize_tag(Expr result, Expr a, Expr b, Expr c, Expr d, Expr e, Exp
...
+/** Expressions tagged with this intrinsic are considered to be part
+ * of the steady state of some loop with a nasty beginning and end
+ * (e.g. a boundary condition). When Halide encounters likely
+ * intrinsics, it splits the containing loop body into three, and
+ * tries to simplify down all conditions that lead to the likely. For
+ * example, given the expression: select(x < 1, bar, x > 10, bar,
+ * likely(foo)), Halide will split the loop over x into portions where
+ * x < 1, 1 <= x <= 10, and x > 10.
+ *
+ * You're unlikely to want to call this directly. You probably want to
```
Comment (steven-johnson) [inline src/IROperator.h]: select(!using_boundary_conditions, use_likely, likely(not_use_likely))

Comment (kree-colemcalughlin) [inline src/IROperator.h]: This is an interesting point. If someone tags the wrong branch as likely things could go horribly wrong, but I guess that is like saying if someone chooses a really bad schedule things will be really slow. But should we hide the use of likely from the outside world more explicitly?

File: src/PartitionLoops.cpp (hunk 1 above)
Comment (abadams) [inline src/PartitionLoops.cpp]: Oops, old code left in. The line below does the simplification.

File: src/PartitionLoops.cpp (hunk 2 above)
Comment (abadams) [inline src/PartitionLoops.cpp]: Yeah, I was worried about explosion. 9 cases already seemed to large to me, so I limited it to one level. The downside is that if you have a schedule like this:

f.vectorize(x, 8).reorder(y, x);

Then it'll optimize away any boundary conditions in y instead of x, which is probably not what you want. I think these schedules are pretty rare though. I only ever do that if I'm also parallelizing over x, which would partially disable the optimization anyway.

File: src/Solve.cpp (hunk 3 above)
Comment (abadams) [inline src/Solve.cpp]: Calling mutate on the expression from the external scope will put a solved version of it in the cache on first use, and pull it out of the cache on subsequent uses.

File: src/BoundaryConditions.cpp (hunk 4 above)
Comment (abadams) [inline src/BoundaryConditions.cpp]: constant_exterior now uses repeat_edge at its core.

**Comment by abadams** [comment]
Throwing clamps on the input will still do the right thing too. For more complex boundary conditions you definitely now need to use the boundary condition helpers. I'll update some comments.

File: src/PartitionLoops.cpp [hunk 6]
```diff
@@ -0,0 +1,924 @@
...
+            //
+            // Note that the bounds we use are actually slightly
+            // conservative. E.g. we claim the prologue can run up
+            // till min_steady, where it actually runs to
+            // clamp(min_steady, loop_min, loop_max). Fortunately, the
+            // case where our bound is incorrect is the case where the
+            // prologue or epilogue have extent zero, so the incorrect
+            // simplification doesn't matter.
+            BoundsBasedSimplify simplify_prologue(op->name, op->min, min_steady.defined() ? min_steady - 1 : Expr());
+            BoundsBasedSimplify simplify_epilogue(op->name, max_steady, op->min + op->extent - 1);
```
Comment (kree-colemcalughlin) [inline src/PartitionLoops.cpp]: It looks like these objects are no longer used.

Comment (abadams) [inline src/PartitionLoops.cpp]: Oh wow, I forgot to delete a bunch of dead code. Good catch.

File: src/PartitionLoops.cpp (hunk 2 above)
Comment (kree-colemcalughlin) [inline src/PartitionLoops.cpp]: One idea I had was to mutate the outer loops first and move inward, but only recurse on the steady state. So instead of growing as 3^n I think the number of loops would grow as 2*n + 1. Do you think that might work?

Comment (abadams) [inline src/PartitionLoops.cpp]: Cool idea - I just tried that out. It does indeed fully simplify the inner loop and make five copies. However, for the boundary conditions performance test things actually got slightly worse. Maybe the expansion in code size? I don't think the boundary condition handling for the y coord costs anything there, because it's just a matter of setting up pointers to the right scanline outside of the inner loop. 

I'll make a note of this in the source. If we hit use cases where we need to handle more than the innermost loop, we can reevaluate.

**Comment by abadams** [comment]
Ok, I think I've addressed all the comments. Works on windows too (the new windows build-bot is finally working!)
